aws/lambda/servicelab-ingestor/lambda_function.py

```diff
@@ -0,0 +1,155 @@
+)
+
+METADATA_REGEX = re.compile(
+    r"pytorch/benchmarks/dynamo/manifold/(?P<experiment_id>\d+)/(?P<trial_id>\d+)/(?P<compiler>\w+)-(?P<model>\w+)-(?P<mode>\w+)-(?P<benchmark_type>\w+)-\w+\.\w+\.\d+_(?P<retry>\d+)\.(?P<experiment_type>\w+)-\w+\.csv"
+)
+
+
+class EventType(Enum):
+    PUT = "ObjectCreated"
+
+
+def lambda_handler(event: Any, context: Any) -> None:
+    counts = defaultdict(int)
+    for record in event["Records"]:
+        event_name = record.get("eventName", "")
+        try:
+            if event_name.startswith(EventType.PUT.value):
+                upsert_document(record)
+            else:
+                warn(f"Unrecognized event type {event_name} in {json.dumps(record)}")
+
+            counts[event_name] += 1
+        except Exception as error:
+            warn(f"Failed to process {json.dumps(record)}: {error}")
+
+    print(f"Finish processing {json.dumps(counts)}")
+
+
+def extract_bucket(record: Any) -> Optional[str]:
+    return record.get("s3", {}).get("bucket", {}).get("name", None)
+
+
+def extract_key(record: Any) -> Optional[str]:
+    return record.get("s3", {}).get("object", {}).get("key", None)
+
+
+def extract_metadata(record: Any) -> Dict[str, Any]:
+    key = extract_key(record)
+    m = re.match(METADATA_REGEX, key)
+    if not m:
+        return {}
+
+    return {
+        "id": key.replace(".csv", ""),
+        "servicelab_experiment_id": m["experiment_id"],
+        "servicelab_experiment_type": m["experiment_type"],
+        "servicelab_trial_id": m["trial_id"],
+        "epoch_timestamp": int(
+            parser.parse(record.get("eventTime")).timestamp() * 1000
+        ),
+        "compiler": m["compiler"],
+        "mode": m["mode"],
+        # TODO (huydhn): Only TorchBench models are supported for now, not HF or TIMM
+        "suite": "torchbench",
+        # TODO (huydhn): Figure out a way to not hardcode this field
+        "dtype": "bfloat16" if m["mode"] == "inference" else "amp",
+        "benchmark_type": m["benchmark_type"],
+    }
+
+
+def read_csv(bucket: str, key: str):
+    response = S3_CLIENT.get_object(Bucket=bucket, Key=key)
+    for r in csv.DictReader(
+        response["Body"].read().decode("utf-8").split("\n"), delimiter=","
+    ):
+        yield r
+
+
+def upsert_document(record: Any) -> None:
+    """
+    Insert a new doc or modify an existing document. Note that ClickHouse doesn't really
+    update the document in place, but rather adding a new record for the update
+    """
+    bucket, key = extract_bucket(record), extract_key(record)
+    if not bucket or not key:
+        return
+
+    metadata = extract_metadata(record)
+    if not metadata:
+        return
+
+    count = 0
+    body_str = ""
+    for r in read_csv(bucket, key):
+        count += 1
+
+        # Populate the record metadata
+        r.update(metadata)
+        body_str += json.dumps(r) + "\n"
+
+    if body_str:
+        print(f"UPSERTING {count} records into {CLICKHOUSE_TABLE}")
```

Comment:
The query has INSERT, but the comment says UPSERT.
Is this intentional?

Reply:
Yeah, it should just be insert in this case as there is no updating existing records.  Let me update the comment.  Just FYI, ClickHouse doesn't really update anything from what I have learnt so far, it appends a new record.  So updating and inserting are basically the same with the same sort key.